src/core/utils/hooks/useOverflow.tsx

```diff
@@ -33,22 +35,30 @@ export const useOverflow = <T extends HTMLElement>(
 ) => {
   const containerRef = React.useRef<T>(null);
 
-  const [visibleCount, setVisibleCount] = React.useState(items.length);
-  const overflowBreakpoints = React.useRef<number[]>([]);
+  const [visibleCount, setVisibleCount] = React.useState(() =>
+    disabled ? items.length : Math.min(items.length, STARTING_MAX_ITEMS_COUNT),
+  );
 
-  React.useLayoutEffect(() => {
-    setVisibleCount(items.length);
-    overflowBreakpoints.current = [];
-  }, [items]);
+  const needsFullRerender = React.useRef(true);
 
   const [containerWidth, setContainerWidth] = React.useState<number>(0);
+  const previousContainerWidth = React.useRef<number>(0);
   const updateContainerWidth = React.useCallback(
     ({ width }) => setContainerWidth(width),
     [],
   );
   const [resizeRef, observer] = useResizeObserver<T>(updateContainerWidth);
   const resizeObserverRef = React.useRef(observer);
 
+  React.useLayoutEffect(() => {
+    if (disabled) {
+      setVisibleCount(items.length);
+    } else {
+      setVisibleCount(Math.min(items.length, STARTING_MAX_ITEMS_COUNT));
+      needsFullRerender.current = true;
+    }
+  }, [containerWidth, disabled, items]);
+
   const mergedRefs = useMergedRefs(containerRef, resizeRef);
```

Comment:
Might be more efficient if we don't merge `resizeRef` when `disabled` is true. Ref wont be attached and ResizeObserver won't be created